virtual_rainforest/soil/__init__.py

```diff
@@ -0,0 +1,69 @@
+import json
+from pathlib import Path
+from typing import Any
+
+from numpy import datetime64, timedelta64
+
+from virtual_rainforest.core.config import register_schema
+from virtual_rainforest.core.logger import LOGGER, log_and_raise
+from virtual_rainforest.core.model import register_model
+from virtual_rainforest.soil.model import InitialisationError, SoilModel
+
+
+@register_schema("soil")
+def schema() -> dict:
+    """Defines the schema that the soil module configuration should conform to."""
+
+    schema_file = Path(__file__).parent.resolve() / "soil_schema.json"
+
+    with schema_file.open() as f:
+        config_schema = json.load(f)
+
+    return config_schema
+
+
+@register_model("soil")
+def generate_soil_model(
```

Comment:
This is generally what is called a `factory` method of a class: a class method that takes a configuration - often read from disk - and creates the object of that class. 

So, if used, this function will become a class method of the main class below:

```python
class SoilModel(BaseModel):

    @classmethod
    def factory(cls, config: dict[str, Any]) -> SoilModel:
        # Same contents than now, but ending in:
        return cls(start_time, end_time, update_interval, no_layers)
```

If used, `factory` should become an abstract method (and a class method) of `BaseClass` so all subclasses need to implement their own version.

If used in combination with `__init_subclass__` above, then `register_model` is not needed, and the only difference if that, when initialising the model from the registry, it should use:

```python
MODEL_REGISTRY[`soil`].factory(config)
```

All models will be initialised in the same way, with the `factory` method.

I think it is neater to keep the factory alongside the class, but it is up to you.